python/ray/tune/tests/test_tune_save_restore.py

```diff
@@ -0,0 +1,160 @@
+            config={
+                "env": "CartPole-v0",
+                "log_level": "DEBUG"
+            })
+
+        exp_dir = os.path.join(absolute_local_dir, exp_name)
+        _, abs_trial_dir = self._get_trial_dir(exp_dir)
+
+        self.assertIsNone(trial.error_file)
+        self.assertEqual(trial.local_dir, exp_dir)
+        self.assertEqual(trial.logdir, abs_trial_dir)
+
+        self.assertTrue(os.path.isdir(absolute_local_dir), absolute_local_dir)
+        self.assertTrue(os.path.isdir(exp_dir))
+        self.assertTrue(os.path.isdir(abs_trial_dir))
+        self.assertTrue(
+            os.path.isfile(
+                os.path.join(abs_trial_dir, "checkpoint_1/checkpoint-1")))
+
+    def _restore(self, exp_name, local_dir, absolute_local_dir):
+        trial_name, abs_trial_dir = self._get_trial_dir(
+            os.path.join(absolute_local_dir, exp_name))
+
+        checkpoint_path = os.path.join(
+            local_dir, exp_name, trial_name,
+            "checkpoint_1/checkpoint-1")  # Relative checkpoint path
+
+        # The file tune would find. The absolute checkpoint path.
+        tune_find_file = os.path.abspath(os.path.expanduser(checkpoint_path))
+        self.assertTrue(
+            os.path.isfile(tune_find_file),
+            "{} is not exist!".format(tune_find_file))
+
+        trial, = tune.run(
+            self.MockTrainable,
+            name=exp_name,
+            stop={"training_iteration": 2},  # train one more iteration.
+            restore=checkpoint_path,  # Restore the checkpoint
+            config={
+                "env": "CartPole-v0",
+                "log_level": "DEBUG"
+            })
+        self.assertIsNone(trial.error_file)
+
+    def testDottedRelativePath(self):
+        local_dir = "./test_dotted_relative_local_dir"
+        exp_name = self.prefix + "DottedRelativeLocalDir"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
+        self._restore(exp_name, local_dir, absolute_local_dir)
+        shutil.rmtree(absolute_local_dir, ignore_errors=True)
+
+    def testRelativePath(self):
+        local_dir = "test_relative_local_dir"
+        exp_name = self.prefix + "RelativePath"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
+        self._restore(exp_name, local_dir, absolute_local_dir)
+        shutil.rmtree(absolute_local_dir, ignore_errors=True)
+
+    def testTildeAbsolutePath(self):
+        local_dir = "~/test_tilde_absolute_local_dir"
+        exp_name = self.prefix + "TildeAbsolutePath"
+        absolute_local_dir = os.path.expanduser(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
+        self._restore(exp_name, local_dir, absolute_local_dir)
+        shutil.rmtree(absolute_local_dir, ignore_errors=True)
+
+    def testAbsolutePath(self):
+        local_dir = "~/test_absolute_local_dir"
```

Comment:
```suggestion
        local_dir = "~/test_absolute_local_dir"
        self.assertFalse(os.path.exists(local_dir))
```

Reply:
A failed run of `test_tune_save_restore.py` may left this dir `~/test_absolute_local_dir` existing. Therefore I remove the potentially existing dir before training, otherwise you have to delete it by hand.

```
    def _train(self, exp_name, local_dir, absolute_local_dir):
	        shutil.rmtree(absolute_local_dir, ignore_errors=True)
```

Reply:
I think it would be better to handle the failure cleanly to remove the directory, and then check the non-existence of it. This just because, I think deleting directories from users' home directory, even with hard-coded names like this, is kinda precarious.

Reply:
Maybe you can add `self.assertFalse(os.path.exists("~/test_absolute_local_dir"))` in `setUp` and `shutil.rmtree("~/test_absolute_local_dir", ignore_errors=True)` in `tearDown` or something similar?